File: servo_interface/scripts/servo_interface.py
#!/usr/bin/python3
import os
import sys
import time
import rospy
from math import degrees
from trajectory_msgs.msg import JointTrajectory


SERVO_PINS = [15,14,13,  12,11,10,  9,8,7,  6,5,4] #rf lf rb lb
servo_configs = []

class Servo:
    def __init__(self, pin):
        path = f"/sys/class/pwm/pwmchip0/pwm{pin}/duty_cycle"
        self.servo = open(path, 'w')
    
    @staticmethod
    def angle_to_duty_cycle(angle: float):
        return int(round(500_000+1000_000*angle/90))

    def set_duty_cycle(self, duty_cycle: int):
        '''Set the duty cycle for the servo. Range 500_000 .. 2500_000'''
        self.servo.write(str(duty_cycle))
        self.servo.flush()

    def set_angle(self, angle: float):
        self.set_duty_cycle(self.angle_to_duty_cycle(angle))

class ServoInterface:
    def __init__(self):
        self.servos = [Servo(pin) for pin in SERVO_PINS]
        self.count = 0

    def callback(self, data):
        points = data.points[0]
        joint_positions = points.positions
        if self.count == 0:
            print(joint_positions)
        lf1_position = degrees(joint_positions[0])
        lf2_position = degrees(joint_positions[1])
        lf3_position = degrees(joint_positions[2])
        rf1_position = degrees(joint_positions[3])
        rf2_position = degrees(joint_positions[4])
        rf3_position = degrees(joint_positions[5])
        lb1_position = degrees(joint_positions[6])
        lb2_position = degrees(joint_positions[7])
        lb3_position = degrees(joint_positions[8])
        rb1_position = degrees(joint_positions[9])
        rb2_position = degrees(joint_positions[10])
        rb3_position = degrees(joint_positions[11])

        a1 = servo_configs[0]+rf1_position
        a2 = servo_configs[1]-rf2_position
        a3 = servo_configs[2]-90-rf2_position-rf3_position

        a4 = servo_configs[3]+lf1_position
        a5 = servo_configs[4]+lf2_position
        a6 = servo_configs[5]+90+lf2_position+lf3_position

        a7 = servo_configs[6]-rb1_position
        a8 = servo_configs[7]-rb2_position
        a9 = servo_configs[8]-90-rb2_position-rb3_position

        a10 = servo_configs[9]-lb1_position
        a11 = servo_configs[10]+lb2_position
        a12 = servo_configs[11]+90+lb2_position+lb3_position
        
        self.servos[0].set_angle(a1)
        self.servos[1].set_angle(a2)
        self.servos[2].set_angle(a3)

        self.servos[3].set_angle(a4)
        self.servos[4].set_angle(a5)
        self.servos[5].set_angle(a6)

        self.servos[6].set_angle(a7)
        self.servos[7].set_angle(a8)
        self.servos[8].set_angle(a9)

        self.servos[9].set_angle(a10)
        self.servos[10].set_angle(a11)
        self.servos[11].set_angle(a12)

        if self.count == 0:
            print(a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12)
        self.count += 1
        
# def set_servo_angle(pin,angle):
# 	duty_cycle = int(500000+11111.11111*angle)
# 	servo_pin = "/sys/class/pwm/pwmchip0/pwm"+str(pin)+"/duty_cycle"
# 	f = open(servo_pin, "w")
# 	f.write(str(int(duty_cycle)))

def get_param():
    global servo_configs
    servo_configs.append(rospy.get_param('rf1_initial_angle'))
    servo_configs.append(rospy.get_param('rf2_initial_angle'))
    servo_configs.append(rospy.get_param('rf3_initial_angle'))

    servo_configs.append(rospy.get_param('lf1_initial_angle'))
    servo_configs.append(rospy.get_param('lf2_initial_angle'))
    servo_configs.append(rospy.get_param('lf3_initial_angle'))

    servo_configs.append(rospy.get_param('rb1_initial_angle'))
    servo_configs.append(rospy.get_param('rb2_initial_angle'))
    servo_configs.append(rospy.get_param('rb3_initial_angle'))

    servo_configs.append(rospy.get_param('lb1_initial_angle'))
    servo_configs.append(rospy.get_param('lb2_initial_angle'))
    servo_configs.append(rospy.get_param('lb3_initial_angle'))

def listener():
    rospy.init_node('servo_interface',anonymous=True)
    get_param()
    servos = ServoInterface()
    rospy.Subscriber("/joint_group_position_controller/command",JointTrajectory,
        servos.callback,queue_size=1)
    rospy.spin()

if __name__=='__main__':
    listener()

File: servo_interface/src/servo_interface.cpp
#include "ros/ros.h"
#include "trajectory_msgs/JointTrajectory.h"
#include <math.h>
#include <stdio.h>

static double degrees(double rad) {
  return rad * 180 / M_PI;
}

static int angle_to_duty_cycle(double angle) {
  return round(500000 + 1000000 * angle / 90.); 
}

static int SERVO_PINS[] = { 15,14,13,  12,11,10,  9,8,7,  6,5,4 };  // rf lf rb lb
static const size_t NSERVOS = 12;

class ServoInterface {
  // std::vector<double> servo_configs;
  double servo_configs[NSERVOS];
  FILE* servos[NSERVOS];
  ros::Subscriber sub;
  int count;

  public:
    ServoInterface(ros::NodeHandle *nh) {
    sub = nh->subscribe("/joint_group_position_controller/command", 1, &ServoInterface::callback, this);
    const char* keys[NSERVOS] = {
      "rf1_initial_angle",
      "rf2_initial_angle",
      "rf3_initial_angle",
      
      "lf1_initial_angle",
      "lf2_initial_angle",
      "lf3_initial_angle",

      "rb1_initial_angle",
      "rb2_initial_angle",
      "rb3_initial_angle",

      "lb1_initial_angle",
      "lb2_initial_angle",
      "lb3_initial_angle",
    };
    for (int i = 0; i < NSERVOS; i++) {
      double x;
      bool res = nh->getParam(keys[i], x);
      ROS_ASSERT_MSG(res, "Failed to get param %s", keys[i]);
      servo_configs[i] = x;
    }
    for (int i = 0; i < NSERVOS; i++) {
      char path[100];
      int pin = SERVO_PINS[i];
      std::snprintf(path, sizeof(path), "/sys/class/pwm/pwmchip0/pwm%d/duty_cycle", pin);
      FILE *f = fopen(path, "w");
      ROS_ASSERT_MSG(f != NULL, "Failed to open %s", path);
      servos[i] = f;
    }
    count = 0;
  }

  void set_angle(int i, double angle) {
    char buf[100];
    int duty_cycle = angle_to_duty_cycle(angle);
    FILE *servo = servos[i];
    std::fprintf(servo, "%d", duty_cycle);
    std::fflush(servo);
  }

  void callback(const trajectory_msgs::JointTrajectory::ConstPtr& msg) {
    auto positions = msg->points[0].positions;
    double lf1 = degrees(positions[0]);
    double lf2 = degrees(positions[1]);
    double lf3 = degrees(positions[2]);
    double rf1 = degrees(positions[3]);
    double rf2 = degrees(positions[4]);
    double rf3 = degrees(positions[5]);
    double lb1 = degrees(positions[6]);
    double lb2 = degrees(positions[7]);
    double lb3 = degrees(positions[8]);
    double rb1 = degrees(positions[9]);
    double rb2 = degrees(positions[10]);
    double rb3 = degrees(positions[11]);

    double angles[NSERVOS] = {
       rf1,
      -rf2,
      -90 - rf2 - rf3,

      lf1,
      lf2,
      90 + lf2 + lf3,

      -rb1,
      -rb2,
      -90 - rb2 - rb3,

      -lb1,
      lb2,
      90 + lb2 + lb3,
    };
    for (int i = 0; i < NSERVOS; i++) {
      double angle = angles[i] + servo_configs[i];
      set_angle(i, angle);
    }
  }
};


int main(int argc, char **argv)
{
  ros::init(argc, argv, "servo_listener"); 
  ros::NodeHandle nh;
  ServoInterface si(&nh);
  ros::spin();
  return 0;
}